### PR TITLE
Fix #1801: Store the already processed byte string in Val.Chars

### LINKED ISSUE #1801: Val.Chars() should store the processed string literal, not the Scala syntax
See https://github.com/scala-native/scala-native/pull/1744#issuecomment-629223184

### PR DESCRIPTION
**Pull request by WojciechMazur**
Resolves #1801 
This PR may break NIR compability. I think there is no need to updated versioning due to other breaking changes which already updated version, since there was no release in between. 
 
Each string in Val.Chars is already escaped literal

### CODE REVIEW

**Comment by LeeTibbert** [comment]
Wojciech
Please pardon me if I am missing something.    I was curious if this PR also fixed 
Issue #1737, so I cloned & built your branch and ran it with some unit-test cases from
PR #1744.   
```
 test("""the value of c"..." literals""") {
    assertEquals("\t", fromCString(c"\t"))
    assertEquals("\\t", fromCString(c"\\t"))
    assertEquals("\u0020\u0020\u0061\u0062", fromCString(c"\040\40\141\x62"))

//    assertEquals("\\n", fromCString(c"\\n"))                                 
    assertEquals("\n", fromCString(c"\n"))
//    assertEquals("\\r", fromCString(c"\\r"))                                 
    assertEquals("\r", fromCString(c"\r"))
    assertEquals("\u0020\u0020\u0061\u0062", fromCString(c"\040\40\141\x62"))

    // assertEquals("""                                                        
    // {                                                                       
    //   "greeting": "Hello world!"                                            
    // }""",                                                                   
    //              fromCString(c"""                                           
    // {                                                                       
    //   "greeting": "Hello world!"                                            
    // }"""))                                                                  
  }
```
 The Hello World case fails, which is fair enough and answers my motivating question.
However, the two other commented out cases fail. It looks to me like they should work.

Thank you for considering this (and thank you for the underlying PR.)

**Comment by WojciechMazur** [comment]
> However, the two other commented out cases fail. It looks to me like they should work.

It's strange since those two cases you mentioned works on my machine. May it be OS related issue? I'm using Fedora 32.
I may dig into triple quote case though.

File: nscplugin/src/main/scala/scala/scalanative/nscplugin/NirGenExpr.scala [hunk 1]
```diff
@@ -1277,7 +1279,8 @@ trait NirGenExpr { self: NirGenPhase =>
                               _))))))),
             _),
             _) =>
-          val chars = Val.Chars(str.replace("\\n", "\n").replace("\\r", "\r"))
+
+          val chars = Val.Chars(treatEscapes(str))
```
Comment (sjrd) [inline nscplugin/src/main/scala/scala/scalanative/nscplugin/NirGenExpr.scala:1283]: Is there any reason why you're not using the built-in `StringContext.treatEscapes(str)`?

Comment (WojciechMazur) [reply nscplugin/src/main/scala/scala/scalanative/nscplugin/NirGenExpr.scala:1283]: It would throw exceptions for unknown escapes from C, that's why it could not be used.

Comment (sjrd) [reply nscplugin/src/main/scala/scala/scalanative/nscplugin/NirGenExpr.scala:1283]: Why would we want to support escapes from C? This is a Scala string interpolation, it should respect Scala's string interpolation rules.

File: tools/src/main/scala/scala/scalanative/codegen/CodeGen.scala [hunk 2]
```diff
@@ -534,52 +534,10 @@ object CodeGen {
...
-                  else "\\" + hex
-                }
-                loop(idx + 2 + hex.length)
-              case unknown =>
-                // clang warns but allows unknown escape sequences, while java emits errors
-                str(unknown); loop(idx + 2)
-            }
-        }
-      loop(0)
+      str(value.replace("\\", "\\" * 2))
```
Comment (sjrd) [inline tools/src/main/scala/scala/scalanative/codegen/CodeGen.scala:540]: We also have to escape other characters that would be treated specially by LLVM's parser. At the very least, the character `"` must be escaped, but I'm sure other characters need escaping as well. New lines are likely candidates, as well as all control characters. It's unclear from the LLVM docs what the possible escape formats are, however it's clear that double-hex-digit bytes are allowed, so we should use that for all characters that aren't safe.

File: nir/src/main/scala/scala/scalanative/nir/Vals.scala [hunk 3]
```diff
@@ -175,59 +175,7 @@ object Val {
   final case class StructValue(values: Seq[Val])                  extends Val
   final case class ArrayValue(elemty: nir.Type, values: Seq[Val]) extends Val
   final case class Chars(value: java.lang.String) extends Val {
```
Comment (sjrd) [inline nir/src/main/scala/scala/scalanative/nir/Vals.scala:177]: Since `c""` strings are `CString`s, which are `Ptr[CChar] = Ptr[Byte]`, we should directly store `value` as a `Seq[Byte]` at this point. The compiler plugin will then have to encode a `String` as a sequence of `Byte`s using UTF-8.

**Comment by LeeTibbert** [comment]
Re: failing test cases.  I'll probe around at my end to see if something is wonky in my
environment.   I did a second debug round and was able to replicate my result.

Went one step further and it has something to do with the c interpolation.  I think
I have been hosed by that before. Something about the c interpolate seeing &
translating the \n earlier mumble, mumble.    Let me refresh & clear my memory.
mumble, mumble is not good enough for action.

In particular, I want to see what happens with those test cases with both current
SN mainline and 0.4.0-M2.

I have some things I need to accomplish, like food in the house before pending
lockdown, so it probably be tomorrow, Friday, before I have anything useful.

**Comment by LeeTibbert** [comment]
Update:

##### Short story
 A __nasty__ and astonishing bug should be fixed once NirGenExpr.scala line 1283
is changed to use the recommended idiom `StringContext.treatEscapes(str)`.

Fixing this bug results in a reduction in astonishment that should be prominently
mentioned in the release notes.

One would expect `toCString("\\n")` == `c"\\n`. Currently & historically this is
surprisingly not true.  People in the wild probably have code to workaround this
bug, so there should be a prominent mention of the fix in the release notes.

##### Long story

I have been gored by this bug several times over the past few years but have
never defined it well enough to track down.  I have always mumbled a few
words I can not repeat here,  written it off to 
"Oh! quoting code is always quirky!",  and then coded around it.

On my system, using both SN 0.4.0-M2 and the current commit of this PR,
both of the following definitely hard fail. I believe the same would be true
on Travis CI.

```
assertEquals("\\n", fromCString(c"\\n"))
assertEquals("\\r", fromCString(c"\\r"))
```
Whereas `assertEquals("\\t", fromCString(c"\\t")) passes.

The culprit appears to be the  0.4.0-M2  code at NirGenExpr.scala line 1282
` val chars = Val.Chars(str.replace("\\n", "\n").replace("\\r", "\r"))`. The code in this
PR at 1283 carries that bug forward in its private implementation of treatEscapes().

When I try the block below in scastie, I get the expected 2 characters.
```
val	left = "\\n"
val sc = StringContext.treatEscapes(left)
System.err.printf("|" + sc + "|" + s" len: $sc.length")
```
If authorship issues can be worked out with @lolgab, I recommend that those
two test cases be added to CStringSuite.scala for this PR.


##### The obligatory complication.

Scastie reports that `StringContext.treatEscapes` has been deprecated in Scala 2.13.0
and recommends `processEscapes`.  The latter is documented to disallow octal escape
sequences.  Doing the obvious and using `processEscapes` means that toCString and
the c interpolater will handle octal escapes differently. The former will pass them, the
latter will not.   

toCString() does not mumbleEscapes() the string it is passed. It converts it to bytes
and copies it pretty straightforwardly to the memory it allocates. (For some reason, it
uses an open coded loop rather than a memcpy or some form of array copy. Perhaps
optimized for one or byte strings????),

File: nscplugin/src/main/scala/scala/scalanative/nscplugin/NirGenExpr.scala (hunk 1 above)
Comment (WojciechMazur) [reply nscplugin/src/main/scala/scala/scalanative/nscplugin/NirGenExpr.scala]: I've changed it to use `StringContext.processEscapes` as mentioned earlier

File: tools/src/main/scala/scala/scalanative/codegen/CodeGen.scala (hunk 2 above)
Comment (WojciechMazur) [reply tools/src/main/scala/scala/scalanative/codegen/CodeGen.scala]: All characters other then digits and letters would not be encoded as double-hex-digit (only `\` is encoded as `\\`). One of the problems I've noticed was that LLVM treats `\"'` as two characters, which means it had to be encoded as `\22` anyway. 
I'm only wondering does encoding all whitespaces as hex values is good solution. It's the safest option  spaces or tabs are currently also hex values. Maybe I should have change that?

Comment (sjrd) [reply tools/src/main/scala/scala/scalanative/codegen/CodeGen.scala]: I suggest to encode:

* Every non-ASCII character (> 0x7F)
* All control characters (< 0x20 and also 0x7F), so that includes tab but not space (which is = 0x20)
* `"` itself, obviously

Concretely, if `c < 0x20 || c == '"' || c >= 0x7f`, then encode it, otherwise don't encode it.

And I agree that we should always encode using the format `\xx` where `xx` is a 2-digit hex value.

File: docs/user/interop.rst [hunk 4]
```diff
@@ -357,6 +357,9 @@ strings (similarly to C):
     val msg: CString = c"Hello, world!"
     stdio.printf(msg)
 
+It does not allow any octal values or escape characters not supported by Scala compiler, like ``\a`` or ``\?``.
+It is possible to use C-style hex values up to value 0xFF, eg. ``c"Hello \x61\x62\x63"``
+
```
Comment (LeeTibbert) [inline docs/user/interop.rst:362]: Well said!  I like the examples, they clarify.

I hope that you are not getting frustrated by the review cycles.

Whack a mole time!    I can create an independent issue for the concern below
so that the current PR can proceed. 

I think this section exposes a weakness in the section which follows it. 

Absent any documentation, it is reasonable to believe that the c interpolator
and toCString() have the same results.  One knows that _someone_ is going
to do toCString() and report that it allows escapes which the quote-equivalent-unquote
interpolator does not.
```
Additionally, we also expose two helper functions unsafe.toCString and unsafe.fromCString to convert between C-style and Java-style strings.
```
I think that creating some test cases to figure out what toCString() is actually
doing with regards to escapes and documenting it in this .rst is worthwhile.
I _think_ it is converting from Java strings to bytes and then copying those bytes to allocated memory.  The creation of the Java string would have done escape
processing but toCString() would not. Thus one would need a Java "\\a" to pass
 the two bytes '\' & 'a' to C.  

The same consideration and need for explicit test cases applies to fromCString().
For the sake of this discussion, consider a byte array read in from user input.
I believe that each byte gets visited one by one and converted to a Java 16 bit Character without escape processing.  Thus a CString holding the two bytes
'\' and 'n' become two Java Characters, not one newline.  

I propose that one or more test cases be created and the behavior documented
in this section of the .rst.

File: unit-tests/src/test/scala/scala/scalanative/unsafe/CStringSuite.scala [hunk 5]
```diff
@@ -11,29 +10,52 @@ object CStringSuite extends tests.Suite {
...
+
+    assertEquals("\f", fromCString(c"\f"))
+    assertEquals("\\f", fromCString(c"\\f"))
+
     assertEquals("\t", fromCString(c"\t"))
     assertEquals("\\t", fromCString(c"\\t"))
-    assertEquals("\u0020\u0020\u0061\u0062", fromCString(c"\040\40\141\x62"))
-    assertEquals("\\\\v\t\\abc", fromCString(c"\\\v\t\\abc"))
+
+    assertEquals("\\n", fromCString(c"\\n"))
```
Comment (LeeTibbert) [inline unit-tests/src/test/scala/scala/scalanative/unsafe/CStringSuite.scala:40]: Some people are never happy.  The added cases for \n and \r are helpful.
In super-picky mode, I did notice that the first few sets test the single backslash
case first and the double backslash case second.  Here \n & \r are done the
other way around. If this is the biggest concern I can muster, life is good,
far better than before you put this effort in.

File: unit-tests/src/test/scala/scala/scalanative/unsafe/CStringSuite.scala [hunk 6]
```diff
@@ -11,29 +10,52 @@ object CStringSuite extends tests.Suite {
...
+    assertEquals("\n", fromCString(c"\n"))
+
+    assertEquals("\\r", fromCString(c"\\r"))
+    assertEquals("\r", fromCString(c"\r"))
+
+    assertEquals("""
```
Comment (LeeTibbert) [inline unit-tests/src/test/scala/scala/scalanative/unsafe/CStringSuite.scala:46]: @lolgab will be happy that you fixed this. Me too!  

What do you do with your pile of dead bugs?

File: unit-tests/src/test/scala/scala/scalanative/unsafe/CStringSuite.scala (hunk 5 above)
Comment (LeeTibbert) [reply unit-tests/src/test/scala/scala/scalanative/unsafe/CStringSuite.scala:40]: Looks like I introduced the \n, \\n case order inconsistency  here, or at least lead you down 
the garden path.

Mea culpa! & sorry.

File: docs/user/interop.rst (hunk 4 above)
Comment (sjrd) [reply docs/user/interop.rst:362]: The key insight is that:

* A `String` is a UCS string, i.e., a sequence of 2-byte Chars. It is *usually* interpreted as UTF-16.
* A `CString` is a *byte string*, i.e., a sequence of Bytes. It is *usually* interpreted as UTF-8, but it can also be ASCII or latin1 or whatever.
* The `c"..."` interpolator describes a *byte string*, without any specific encoding. Because the source code is text, *some* encoding needs to be chosen when encoding the text *at compile-time* into a byte string. The natural choice is latin1 aka ISO-8859-1, because it uses 1-byte code points, and its valid code point range (0-255) corresponds to the same range in UTF-16 `Char`s and in abstract Unicode code point.
* `fromCString` and `toCString` are charset-aware. They will always assume that the `String` is UTF-16 (as do the `java.nio.charset` APIs) and take an `Charset` parameter to know what encoding to assume for the byte string. If not specified, it is UTF-8. They are very similar to `String.getBytes` and `new String(Array[Byte])`, except that they work with a `CString` instead of an `Array[Byte]`.

`fromCString` and `toCString` don't care about escapes; escapes are a source-level programming language concern, not a run-time concern. They don't copy byte-by-byte because that would not respect the charset-aware decoding and encoding.

File: unit-tests/src/test/scala/scala/scalanative/unsafe/CStringSuite.scala (hunk 6 above)
Comment (lolgab) [reply unit-tests/src/test/scala/scala/scalanative/unsafe/CStringSuite.scala:46]: Thanks for the mention! 😃
Yes, I'm happy, especially now that I don't have to fix my PR because this one supersedes mine 😄 ( didn't have too much time and willpower lately to finish it :( ).

File: nir/src/main/scala/scala/scalanative/nir/serialization/BinaryDeserializer.scala [hunk 7]
```diff
@@ -265,7 +268,7 @@ final class BinaryDeserializer(buffer: ByteBuffer) {
     case T.DoubleVal      => Val.Double(getDouble)
     case T.StructValueVal => Val.StructValue(getVals)
     case T.ArrayValueVal  => Val.ArrayValue(getType, getVals)
-    case T.CharsVal       => Val.Chars(getString)
+    case T.CharsVal       => Val.Chars(getBytes())
```
Comment (sjrd) [inline nir/src/main/scala/scala/scalanative/nir/serialization/BinaryDeserializer.scala:271]: Here we'll need a deserialization hack to be able to read previous versions or the NIR. The idea is that, if we read an .nir file produced by a previous version, we'll do `getString`, and then process the escapes of that string like the previous linker used to do, to produce bytes. And otherwise we just call `getBytes()`.

How do we know whether we're reading a previous version? It's stored in the `Prelude`. Inside `val header = ...`, we call `Prelude.readFrom(), which returns a value of type `Prelude` that contains this information. Currently we throw it away because we don't need it, but now we'll have to store it.

File: nscplugin/src/main/scala/scala/scalanative/nscplugin/NirGenExpr.scala [hunk 8]
```diff
@@ -1277,7 +1279,9 @@ trait NirGenExpr { self: NirGenPhase =>
                               _))))))),
             _),
             _) =>
-          val chars = Val.Chars(str.replace("\\n", "\n").replace("\\r", "\r"))
+
+          val processedString = StringContext.processEscapes(processHexValues(str))
```
Comment (sjrd) [inline nscplugin/src/main/scala/scala/scalanative/nscplugin/NirGenExpr.scala:1283]: Doing one then the other will be problematic for strings like `c"\\x65"`, which should be processed as `\x65`. But if you first call `processHexValues` you get `\A` before `processEscapes` can realize that the `\\` was an escape for `\` rather than a hex escape. Doing `processEscapes` first would cause similar issues.

I'm afraid there is no other way than dealing with everything at once.

File: tools/src/test/scala/scala/scalanative/NIRCompiler.scala [hunk 9]
```diff
@@ -35,7 +35,7 @@ object NIRCompiler {
   def getCompiler(): api.NIRCompiler = {
     val clazz =
       classLoader.loadClass("scala.scalanative.NIRCompiler")
-    clazz.newInstance match {
+    clazz.getConstructor().newInstance() match {
```
Comment (sjrd) [inline tools/src/test/scala/scala/scalanative/NIRCompiler.scala:38]: This change should go in a separate PR that does only that.

File: nir/src/main/scala/scala/scalanative/nir/serialization/BinaryDeserializer.scala (hunk 7 above)
Comment (WojciechMazur) [reply nir/src/main/scala/scala/scalanative/nir/serialization/BinaryDeserializer.scala:271]: Hey, 
I'm not really sure is it needed in this place, previously this new `getBytes` method is simply extracted logic from previous version of `getString`. Currently getString is defined as 
```  
private def getString(): String = {
    new String(getBytes(), "UTF-8")
  }
```

altough as I understand inside `CodeGen` we should change current logic to 
```
...
      case v :Val.Chars =>
        if(v.revision < 7) legacyGenChars(v.stringValue) //here we're using previous logic
        else genChars(v.bytes) // we're using new logic which assumes string is escaped
```
Am i correct?

Comment (sjrd) [reply nir/src/main/scala/scala/scalanative/nir/serialization/BinaryDeserializer.scala:271]: We cannot have version-dependent behavior in CodeGen. The problem is that things can get inlined from one place to another, so a `Val.Chars` may be copied over to a `Def` that was originally with a different `revision`.

The last and only moment we can use version-dependent logic is in the deserializer. After that, all the NIR must be assumed to be in the latest version, and be handled as such.

File: util/src/main/scala/scala/scalanative/util/StringUtils.scala [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+import scala.annotation.tailrec
+
+object StringUtils {
+
+  /**
+   * Custom implementation of StringContext.processEscapes which also parses hex values
+   * @param str UTF-8 encoded input string optionally containing literal escapes and hex values
+   * @return UTF-8 representation of escaped ByteString
+   */
+  def processEscapes(str: String): String = {
```
Comment (sjrd) [inline util/src/main/scala/scala/scalanative/util/StringUtils.scala:13]: Since this method really parses a *byte string*, it should directly return an `Array[Byte]`. It should not try to encode the byte string into a String, only to then re-decode that String into a byte string by calling `getBytes("UTF-8")`. As such, it should use an `Array.newBuilder[Byte]` instead of a `StringBuilder`.

This will avoid many indirections, and would allow to parse byte strings that are not valid UTF-8 sequences (but could be valid in other interpretations, which we must not prevent).

File: nscplugin/src/main/scala/scala/scalanative/nscplugin/NirGenExpr.scala [hunk 11]
```diff
@@ -1277,23 +1278,26 @@ trait NirGenExpr { self: NirGenPhase =>
                               _))))))),
             _),
             _) =>
-          val chars = Val.Chars(str.replace("\\n", "\n").replace("\\r", "\r"))
-          val const = Val.Const(chars)
-          buf.box(nir.Rt.BoxedPtr, const, unwind)
+        // format: on
+        val chars = Val.Chars(StringUtils.processEscapes(str))
```
Comment (sjrd) [inline nscplugin/src/main/scala/scala/scalanative/nscplugin/NirGenExpr.scala:1282]: The indentation is wrong here.

In general, there are zillions of formatting changes that are unrelated to this PR in this file. They could have come from a transient use of another version of scalafmt or some misconfiguration. Could you please revert all the changes in this file, then reintroduce this specific change (which is only 1 line)?

Comment (WojciechMazur) [reply nscplugin/src/main/scala/scala/scalanative/nscplugin/NirGenExpr.scala]: Sure, it should be fixed now
